qutebrowser/mainwindow/tabbedbrowser.py

```diff
@@ -659,7 +662,8 @@ def on_current_changed(self, idx):
                                usertypes.KeyMode.passthrough]
         for mode in modes_to_leave:
             modeman.leave(self._win_id, mode, 'tab changed', maybe=True)
-
+            
+        modeman.enter(self._win_id, tab._input_mode, 'restore input mode for tab')
```

Comment:
There are multiple issues with this:

- If the old tab was e.g. in hint/caret/prompt/... mode, you'll end up in that mode again, but without the prompt/hints/etc. leading to undefined behaviour or crashes. You probably only want to re-enter insert/passthrough mode, no?
- It breaks the `tabs.persist_mode_on_change` setting.
- It should probably have its own setting as well. Maybe that should even be a merged `tabs.mode_on_change` setting or so. I can take care of that part (after my exams) if you prefer.

Reply:
I have addressed all the noted issues and will provide patches for these soon. 

Thanks for the nice and insightful feedback!